File: javascript/doublyLinkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.prev = null;
    this.next = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }
  push(data) {
    const newNode = new Node(data);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
  }
  pop() {
    if (this.head == null) {
      console.log("underflow");
      return;
    }
    this.head = this.head.next;
    if (this.head !== null) {
      this.head.prev = null;
    } else {
      this.tail = null; // List is now empty
    }
  }
  print() {
    if (this.head == null) {
      console.log("underflow");
      return;
    }
    let current = this.head;
    while (current !== null) {
      process.stdout.write(current.data.toString());
      if (current.next !== null) {
        process.stdout.write(" -> ");
      }
      current = current.next;
    }
  }
}

const dll = new DoublyLinkedList();

dll.push(10);
dll.push(20);
dll.push(30);

dll.print(); // Should print 10, 20, 30

dll.pop(); // Removes 10 (head)
dll.print(); // Should print 20, 30

dll.pop(); // Removes 20 (head)
dll.print(); // Should print 30

dll.pop(); // Removes 30 (head)
dll.print(); // Should print "underflow"

File: python/doublyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def inputValue(self):
        return int(input("Value: "))

    def push(self):
        value = self.inputValue()
        newNode = Node(value)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode

    def pop(self):
        if self.head == None:
            print("underflow")
            return
        self.head = self.head.next
        self.head.prev = None

    def print(self):
        if self.head == None:
            print("underflow")
            return
        current = self.head
        while current != None:
            print(current.data, " -> ", end=" ")
            current = current.next
        print()

    def menu(self):
        print("1. push")
        print("2. pop")
        print("3. print")
        print("4. exit")


if __name__ == "__main__":
    dll: DoublyLinkedList = DoublyLinkedList()
    while True:
        dll.menu()
        choice = int(input("choice: "))
        if choice == 4:
            print("Thanks for using Doubly Linked List")
            break
        elif choice == 1:
            dll.push()
        elif choice == 2:
            dll.pop()
        elif choice == 3:
            dll.print()
